python/src/spark_rapids_ml/tuning.py

```diff
@@ -0,0 +1,167 @@
+    -------
+    tuple
+        (int, float, subModel), an index into `epm` and the associated metric value.
+    """
+    modelIter = est.fitMultiple(train, epm)
+
+    def singleTask() -> Tuple[int, float, Transformer]:
+        index, model = next(modelIter)
+        if isinstance(model, RandomForestClassificationModel) and model.numClasses > 2:
+            metric = model._transformEvaluate(validation, epm[index])
+        else:
+            # TODO: duplicate evaluator to take extra params from input
+            #  Note: Supporting tuning params in evaluator need update method
+            #  `MetaAlgorithmReadWrite.getAllNestedStages`, make it return
+            #  all nested stages and evaluators
+            metric = eva.evaluate(model.transform(validation, epm[index]))
+        return index, metric, model if collectSubModel else None
+
+    return [singleTask] * len(epm)
+
+
+# Copied from pyspark.ml.tuning
+def _gen_avg_and_std_metrics(
+    metrics_all: List[List[float]],
+) -> Tuple[List[float], List[float]]:
+    avg_metrics = np.mean(metrics_all, axis=0)
+    std_metrics = np.std(metrics_all, axis=0)
+    return list(avg_metrics), list(std_metrics)
+
+
+class CrossValidator(SparkCrossValidator):
+    """K-fold cross validation performs model selection by splitting the dataset into a set of
+    non-overlapping randomly partitioned folds which are used as separate training and test datasets
+    e.g., with k=3 folds, K-fold cross validation will generate 3 (training, test) dataset pairs,
+    each of which uses 2/3 of the data for training and 1/3 for testing. Each fold is used as the
+    test set exactly once.
+
+    It is the gpu version CrossValidator which fits multiple models in a single pass for a single
+    training dataset and transforms/evaluates in a single pass for a single model.
+    """
+
+    def _fit(self, dataset: DataFrame) -> "CrossValidatorModel":
+        est = self.getOrDefault(self.estimator)
+        eva = self.getOrDefault(self.evaluator)
+
+        if not (
+            isinstance(est, RandomForestClassifier)
+            and isinstance(eva, MulticlassClassificationEvaluator)
+        ):
+            return super()._fit(dataset)
+
+        epm = self.getOrDefault(self.estimatorParamMaps)
+        numModels = len(epm)
+        nFolds = self.getOrDefault(self.numFolds)
+        metrics_all = [[0.0] * numModels for i in range(nFolds)]
+
+        pool = ThreadPool(processes=min(self.getParallelism(), numModels))
+        subModels = None
+        collectSubModelsParam = self.getCollectSubModels()
+        if collectSubModelsParam:
+            subModels = [[None for j in range(numModels)] for i in range(nFolds)]
+
+        datasets = self._kFold(dataset)
+        for i in range(nFolds):
+            # TODO, Need to get rid of cache when supporting transform/evaluate
+            # in a single pass for multiple models.
+            validation = datasets[i][1].cache()
+            train = datasets[i][0]
```

Comment:
Is the lack of cache the only difference between superclass and this?   If so, since cache is lazy, maybe train be unpersisted in our override of _parallelFitTasks ?   Then we can avoid overriding _fit.

Reply:
> Is the lack of cache the only difference between superclass and this?

Not exactly, spark-rapids-ml also re-implements `_parallelFitTasks` to call `_transformEvaluate()`.

Reply:
Can we avoid overriding _fit by:

1. redefining _parallelFitTasks in pyspark.ml.tuning (ie. 
     def _parallelFitTasks ... (i.e. def from spark_rapids_ml here)
     import pyspark.ml.tuning
     pyspark.ml.tuning._parallelFitTasks = _parallelFitTasks
2. unpersisting train in our customized _parllelFitTasks

Reply:
Since you will be making more involved changes to _fit later, we can drop this suggestion.